kotlin/src/com/google/idea/blaze/kotlin/sync/importer/BlazeKotlinWorkspaceImporter.java

```diff
@@ -0,0 +1,98 @@
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.idea.blaze.base.ideinfo.TargetIdeInfo;
+import com.google.idea.blaze.base.ideinfo.TargetMap;
+import com.google.idea.blaze.base.model.LibraryKey;
+import com.google.idea.blaze.base.model.primitives.LanguageClass;
+import com.google.idea.blaze.base.model.primitives.WorkspaceRoot;
+import com.google.idea.blaze.base.projectview.ProjectViewSet;
+import com.google.idea.blaze.base.sync.projectview.ProjectViewTargetImportFilter;
+import com.google.idea.blaze.base.targetmaps.TransitiveDependencyMap;
+import com.google.idea.blaze.java.sync.model.BlazeJarLibrary;
+import com.google.idea.blaze.kotlin.sync.model.BlazeKotlinImportResult;
+import com.intellij.openapi.project.Project;
+import org.jetbrains.annotations.NotNull;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Stream;
+
+public class BlazeKotlinWorkspaceImporter {
+    private final TargetMap targetMap;
+    private final ProjectViewTargetImportFilter importFilter;
+
+    public BlazeKotlinWorkspaceImporter(
+            Project project,
+            WorkspaceRoot workspaceRoot,
+            ProjectViewSet projectViewSet,
+            TargetMap targetMap) {
+        this.targetMap = targetMap;
+        importFilter = new ProjectViewTargetImportFilter(project, workspaceRoot, projectViewSet);
+    }
+
+    public BlazeKotlinImportResult importWorkspace() {
+        HashMap<LibraryKey, BlazeJarLibrary> libraries = new HashMap<>();
+        HashMap<TargetIdeInfo, ImmutableList<BlazeJarLibrary>> targetLibraryMap = new HashMap<>();
+
+        collectTransitiveLibsFromKotlinSourceTargets(libraries, targetLibraryMap);
+
+        return new BlazeKotlinImportResult(
+                ImmutableList.copyOf(libraries.values()),
+                ImmutableMap.copyOf(targetLibraryMap));
+    }
+
+    private void collectTransitiveLibsFromKotlinSourceTargets(
+            HashMap<LibraryKey, BlazeJarLibrary> libraries,
+            HashMap<TargetIdeInfo, ImmutableList<BlazeJarLibrary>> targetLibraryMap
+    ) {
+        List<BlazeJarLibrary> scratchLibSet = new ArrayList<>();
+
+        targetMap
+                .targets().stream()
+                .filter(target -> target.kind.languageClass.equals(LanguageClass.KOTLIN))
+                .filter(importFilter::isSourceTarget)
+                .forEach(ideInfo -> withTransitiveTargets(ideInfo).forEach(depIdeInfo -> {
+                    scratchLibSet.clear();
+                    depIdeInfo.javaIdeInfo.jars.stream().map(BlazeJarLibrary::new).forEach(depJar -> {
+                        libraries.putIfAbsent(depJar.key, depJar);
+                        scratchLibSet.add(depJar);
+                    });
+                    if (depIdeInfo.kind.languageClass == LanguageClass.KOTLIN) {
```

Comment:
Just do a `.filter(depIdeInfo.kind.languageClass == LanguageClass.KOTLIN)` before `.forEach`?

Reply:
This would change the semantics. The process does two things:

1. Collect all external transitive libraries reachable from a Kotlin target.
2. Create a lookup cache to be used in setting up the library table for Kotlin.

if we pull the filter up we won't be doing (1) --i.e.,  plain java libraries won't be reachable from Kotlin targets. Is there something in the Java plugin that will take care of this for JavaLikeLanguage ?

Reply:
Never mind my reasoning is wrong and double injects the java libraries (the Scala sync plugin has a comment on this being ok).. So prefiltering is correct.

 I've updated the method body to:

```
    private void collectTransitiveLibsFromKotlinSourceTargets(
            HashMap<LibraryKey, BlazeJarLibrary> libraries,
            HashMap<TargetIdeInfo, ImmutableList<BlazeJarLibrary>> targetLibraryMap
    ) {
        Stream<TargetIdeInfo> transitiveClosureOfKotlinTargets = targetMap
                .targets().stream()
                .filter(target -> target.kind.languageClass.equals(LanguageClass.KOTLIN) && importFilter.isSourceTarget(target))
                .flatMap(this::withTransitiveTargets)
                .filter(depIdeInfo -> depIdeInfo.kind.languageClass == LanguageClass.KOTLIN);

        List<BlazeJarLibrary> scratchLibSet = new ArrayList<>();

        transitiveClosureOfKotlinTargets.forEach(depIdeInfo -> {
            scratchLibSet.clear();
            // noinspection ConstantConditions javaIdeInfo has been asserted in ::withTransitiveTargets
            depIdeInfo.javaIdeInfo.jars.stream().map(BlazeJarLibrary::new)
                    .forEach(depJar -> {
                libraries.putIfAbsent(depJar.key, depJar);
                scratchLibSet.add(depJar);
            });
            targetLibraryMap.put(depIdeInfo, ImmutableList.copyOf(scratchLibSet));
        });
    }
```

Reply:
I'm not a fan of the `scratchLibSet`. They way it's cleared and built seems prone to bugs. Does this work?
```
    targetMap
        .targets()
        .stream()
        .filter(target -> target.kind.languageClass.equals(LanguageClass.KOTLIN))
        .filter(importFilter::isSourceTarget)
        .map(target -> TransitiveDependencyMap.getTransitiveDependencies(target.key, targetMap))
        .flatMap(Collection::stream)
        .map(targetMap::get)
        .filter(Objects::nonNull)
        .filter(info -> info.kind.languageClass == LanguageClass.KOTLIN)
        .filter(info -> info.javaIdeInfo != null)
        .forEach(
            info -> {
              List<BlazeJarLibrary> targetLibraries =
                  info.javaIdeInfo
                      .jars
                      .stream()
                      .map(BlazeJarLibrary::new)
                      .peek(jar -> libraries.putIfAbsent(jar.key, jar))
                      .collect(Collectors.toList());
              targetLibraryMap.put(info, ImmutableList.copyOf(targetLibraries));
```

Reply:
Interesting, I've never used peek ! The huge process chain will be confusing to interpret later on, but I have removed the tally.